### PR TITLE
Set OcpMaxVersion property in merge-index-image request in IIB

### PR DESCRIPTION
**Pull request by yashvardhannanavati**
Refers to CLOUDDST-8769

### CODE REVIEW

File: iib/workers/tasks/utils.py [hunk 1]
```diff
@@ -50,6 +121,130 @@ def get_binary_image_from_config(ocp_version, distribution_scope, binary_image_c
...
+            # check if the failure is due to the presence of deprecated APIs
+            # and absence of the 'olm.maxOpenShiftVersion' property
+            # Note: there is no other error in the sdk that mentions this field
+            for msg in output['outputs']:
+                if 'olm.maxOpenShiftVersion' in msg['message']:
+                    return True
+    return False
+
+
+def serve_index_registry(db_path):
```
Comment (pbortlov) [inline iib/workers/tasks/utils.py:163]: @yashvardhannanavati I like your idea of moving code out of build.py but utils.py is growing extremely. Can we make a new file called opm or something like that?

Comment (yashvardhannanavati) [reply iib/workers/tasks/utils.py:163]: A new file simply for these functions only wouldn't make sense IMO. We should file an issue and discuss it with the guild before doing a major refactoring

Comment (pbortlov) [reply iib/workers/tasks/utils.py:169]: sounds good :) please ping me in the jira

Comment (yashvardhannanavati) [reply iib/workers/tasks/utils.py:169]: Sounds good. In the meanwhile, could you please help review this PR?

**Review by lipoja — DISMISSED** [review]

**Comment by pbortlov** [comment]
@yashvardhannanavati did you rebase with the master? it seems there are problems with `_skope_copy` which is renamed to `skope_copy`.

File: iib/workers/tasks/utils.py [hunk 2]
```diff
@@ -50,6 +121,136 @@ def get_binary_image_from_config(ocp_version, distribution_scope, binary_image_c
...
+    """
+    Get bundle json from grpcurl response.
+
+    :param list bundles: response from grpcurl call to retrieve list of bundles in
+        an index
+    """
+    return [json.loads(bundle) for bundle in re.split(r'(?<=})\n(?={)', bundles)]
+
+
+def _requires_max_ocp_version(bundle):
```
Comment (pbortlov) [inline iib/workers/tasks/utils.py:134]: Could you move this in front of the function ` add_max_ocp_version_property`? It should be defined before the function is used.

File: iib/workers/tasks/utils.py [hunk 3]
```diff
@@ -50,6 +121,136 @@ def get_binary_image_from_config(ocp_version, distribution_scope, binary_image_c
...
+            # check if the failure is due to the presence of deprecated APIs
+            # and absence of the 'olm.maxOpenShiftVersion' property
+            # Note: there is no other error in the sdk that mentions this field
+            for msg in output['outputs']:
+                if 'olm.maxOpenShiftVersion' in msg['message']:
+                    return True
+    return False
+
+
+def serve_index_registry(db_path):
```
Comment (pbortlov) [inline iib/workers/tasks/utils.py:169]: Could you move this in front of the function `add_max_ocp_version_property`? It should be defined before the function is used.

File: iib/workers/tasks/utils.py [hunk 4]
```diff
@@ -50,6 +121,136 @@ def get_binary_image_from_config(ocp_version, distribution_scope, binary_image_c
...
+        except AddressAlreadyInUse:
+            log.info('Port %d is in use, trying another.', port)
+
+    err_msg = f'No free port has been found after {conf.get("iib_grpc_max_port_tries")} attempts.'
+    log.error(err_msg)
+    raise IIBError(err_msg)
+
+
+@retry(exceptions=IIBError, tries=2, logger=log)
+def serve_index_registry_at_port(db_path, port, max_tries, wait_time):
```
Comment (pbortlov) [inline iib/workers/tasks/utils.py:202]: Could you move this above the function `serve_index_registry`? It should be defined before the function is used.

**Review by pbortlov — CHANGES_REQUESTED** [review]

File: tests/test_workers/test_tasks/test_build.py [hunk 5]
```diff
@@ -1064,27 +1094,38 @@ def test_get_present_bundles_grpc_not_initialize(
 @mock.patch('time.sleep')
 @mock.patch('subprocess.Popen')
 @mock.patch('iib.workers.tasks.build.run_cmd')
```
[the remark this answers is not in the document]
Comment (yashvardhannanavati) [reply tests/test_workers/test_tasks/test_build.py:1096]: Both of them are required.

File: iib/workers/tasks/utils.py (hunk 2 above)
Comment (yashvardhannanavati) [reply iib/workers/tasks/utils.py:134]: May I ask why? I don't think we need to do that in Python unless I am missing something

Comment (pbortlov) [reply iib/workers/tasks/utils.py:134]: I would say best practices and my editor complains a lot :smile:

**Comment by yashvardhannanavati** [comment]
@pbortlov By pulling it locally, and running a merge-index-image request in which, the target_index has more bundles that the source_index. Then in the logs, you can make sure that this piece of code is called when those extra bundles are being added.

File: iib/workers/tasks/utils.py [hunk 6]
```diff
@@ -19,14 +21,83 @@
...
+    port, rpc_proc = serve_index_registry(db_path)
+
+    raw_bundles = run_cmd(
+        ['grpcurl', '-plaintext', f'localhost:{port}', 'api.Registry/ListBundles'],
+        exc_msg='Failed to get bundle data from index image',
+    )
+    rpc_proc.kill()
+
+    # Get bundle json for bundles in the request
+    updated_bundles = list(
```
Comment (pbortlov) [inline iib/workers/tasks/utils.py:83]: Could you update a comment here?
`Get a bundlePath from raw_bundles and check the presence in resolved_bundles to return a list of present bundles`
I am lost here. You convert `raw_bundles` to JSON and then you check presence with resolved_bundles and create a LIST of `updated_bundles` but then you try to work with the `updated_bundles` as a DICTIONARY.

Comment (pbortlov) [reply iib/workers/tasks/utils.py:83]: I check filter() here: https://www.programiz.com/python-programming/methods/built-in/filter

Comment (pbortlov) [reply iib/workers/tasks/utils.py:83]: I have been testing it.

**Pull request merged**